core/state/snapshot/snapshot.go

```diff
@@ -258,6 +258,14 @@ func (t *Tree) Disable() {
 	for _, layer := range t.layers {
 		switch layer := layer.(type) {
 		case *diskLayer:
+
+			layer.lock.RLock()
+			generating := layer.genMarker != nil
+			layer.lock.RUnlock()
```

Comment:
Not sure this is safe. We are iterating the layers here, and locking them individually. What happens if `genMarker` is updated in-between?